itlwm/hal_iwn/ItlIwn.cpp

```diff
@@ -0,0 +1,7306 @@
+ * from power-down sleep mode.
+ */
+void ItlIwn::
+iwn_wakeup_intr(struct iwn_softc *sc)
+{
+    int qid;
+
+    DPRINTF(("ucode wakeup from power-down sleep\n"));
+
+    /* Wakeup RX and TX rings. */
+    IWN_WRITE(sc, IWN_FH_RX_WPTR, sc->rxq.cur & ~7);
+    for (qid = 0; qid < sc->ntxqs; qid++) {
+        struct iwn_tx_ring *ring = &sc->txq[qid];
+        IWN_WRITE(sc, IWN_HBUS_TARG_WRPTR, qid << 8 | ring->cur);
+    }
+}
+
+#ifdef IWN_DEBUG
+/*
+ * Dump the error log of the firmware when a firmware panic occurs.  Although
+ * we can't debug the firmware because it is neither open source nor free, it
+ * can help us to identify certain classes of problems.
+ */
+void ItlIwn::
+iwn_fatal_intr(struct iwn_softc *sc)
+{
+    struct iwn_fw_dump dump;
+    int i;
+
+    /* Check that the error log address is valid. */
+    if (sc->errptr < IWN_FW_DATA_BASE ||
+        sc->errptr + sizeof (dump) >
+        IWN_FW_DATA_BASE + sc->fw_data_maxsz) {
+        XYLog("%s: bad firmware error log address 0x%08x\n",
+            sc->sc_dev.dv_xname, sc->errptr);
+        return;
+    }
+    if (iwn_nic_lock(sc) != 0) {
+        XYLog("%s: could not read firmware error log\n",
+            sc->sc_dev.dv_xname);
+        return;
+    }
+    /* Read firmware error log from SRAM. */
+    iwn_mem_read_region_4(sc, sc->errptr, (uint32_t *)&dump,
+        sizeof (dump) / sizeof (uint32_t));
+    iwn_nic_unlock(sc);
+
+    if (dump.valid == 0) {
+        XYLog("%s: firmware error log is empty\n",
+            sc->sc_dev.dv_xname);
+        return;
+    }
+    XYLog("firmware error log:\n");
+    XYLog("  error type      = \"%s\" (0x%08X)\n",
+        (dump.id < nitems(iwn_fw_errmsg)) ?
+        iwn_fw_errmsg[dump.id] : "UNKNOWN",
+        dump.id);
+    XYLog("  program counter = 0x%08X\n", dump.pc);
+    XYLog("  source line     = 0x%08X\n", dump.src_line);
+    XYLog("  error data      = 0x%08X%08X\n",
+        dump.error_data[0], dump.error_data[1]);
+    XYLog("  branch link     = 0x%08X%08X\n",
+        dump.branch_link[0], dump.branch_link[1]);
+    XYLog("  interrupt link  = 0x%08X%08X\n",
+        dump.interrupt_link[0], dump.interrupt_link[1]);
+    XYLog("  time            = %u\n", dump.time[0]);
+
+    /* Dump driver status (TX and RX rings) while we're here. */
+    XYLog("driver status:\n");
+    for (i = 0; i < sc->ntxqs; i++) {
+        struct iwn_tx_ring *ring = &sc->txq[i];
+        XYLog("  tx ring %2d: qid=%-2d cur=%-3d queued=%-3d\n",
+            i, ring->qid, ring->cur, ring->queued);
+    }
+    XYLog("  rx ring: cur=%d\n", sc->rxq.cur);
+    XYLog("  802.11 state %d\n", sc->sc_ic.ic_state);
+}
+#endif
+
+int ItlIwn::
+iwn_intr(OSObject *object, IOInterruptEventSource* sender, int count)
+{
+    ItlIwn *that = (ItlIwn*)object;
+    struct iwn_softc *sc = &that->com;
+    struct _ifnet *ifp = &sc->sc_ic.ic_if;
+    uint32_t r1, r2, tmp;
+
+    /* Disable interrupts. */
+    IWN_WRITE(sc, IWN_INT_MASK, 0);
```

Comment:
don't need to disable interrupt again.

Reply:
Yeah, just found it was the interrupt was already disabled.